libsql_client/dbapi2/_async_executor.py

```diff
@@ -0,0 +1,213 @@
+            raise RuntimeError("thread already down")
+
+        loop, stop_event = self._control
+
+        async def run_in_main_loop() -> None:
+            self._dbg("notifying _stop_event")
+            stop_event.set()
+
+        self._dbg("thread will stop")
+        coro = run_in_main_loop()
+        future = asyncio.run_coroutine_threadsafe(coro, loop)
+        future.result()
+        self.join()
+        self._control = None
+        self._inf("thread did stop")
+
+    if sys.version_info[:2] >= (3, 9):
+        @overload
+        def submit(
+            self,
+            fn: Callable[P, Awaitable[asyncio.Future[T]]],
+            *args: P.args,
+            **kwargs: P.kwargs,
+        ) -> concurrent.futures.Future[T]:
+            ...
+
+        @overload
+        def submit(
+            self,
+            fn: Callable[P, Awaitable[T]],
+            *args: P.args,
+            **kwargs: P.kwargs,
+        ) -> concurrent.futures.Future[T]:
+            ...
+
+        @overload
+        def submit(
+            self,
+            fn: Callable[P, T],
+            *args: P.args,
+            **kwargs: P.kwargs,
+        ) -> concurrent.futures.Future[T]:
+            ...
+
+    else:
+        @overload
+        def submit(
+            self,
+            fn: Callable[P, Awaitable[asyncio.Future]],
+            *args: P.args,
+            **kwargs: P.kwargs,
+        ) -> concurrent.futures.Future:
+            ...
+
+        @overload
+        def submit(
+            self,
+            fn: Callable[P, Awaitable[T]],
+            *args: P.args,
+            **kwargs: P.kwargs,
+        ) -> concurrent.futures.Future:
+            ...
+
+        @overload
+        def submit(
+            self,
+            fn: Callable[P, T],
+            *args: P.args,
+            **kwargs: P.kwargs,
+        ) -> concurrent.futures.Future:
+            ...
+
+    def submit(
+        self,
+        fn: Callable[P, T],
+        *args: P.args,
+        **kwargs: P.kwargs,
+    ) -> Any:  # variants are covered by overloads
+        if self._control is None:
+            raise RuntimeError("already down")
+
+        async def run_in_main_loop() -> Any:
+            try:
+                self._dbg("calling: %s, args=%s, kwargs=%s", fn, args, kwargs)
+                r = fn(*args, **kwargs)
+                while asyncio.iscoroutine(r) or asyncio.isfuture(r):
+                    r = await r
+                self._dbg(
+                    "finished: %s, args=%s, kwargs=%s, result=%s",
+                    fn, args, kwargs, r,
+                )
+                return r
+            except Exception as e:
+                self._dbg(
+                    "failed: %s, args=%s, kwargs=%s, exc=%s",
+                    fn, args, kwargs, e,
+                    exc_info=e,
+                )
+                raise
+
+        coro = run_in_main_loop()
+        return asyncio.run_coroutine_threadsafe(coro, self._control.loop)
```

Comment:
this is done by asyncio.run_coroutine_threadshafe, it will add a future and handle the cases when it should fail.

However, given the synchronous usage and the rest of the code, this can't happen, the connection close will send the termination and dispose the object right away

Reply:
not sure I get what you mean, you mean 2 threads calling close() at the same time? If so I tested and you get `RuntimeError: Event loop is closed`.

I don't think it's worth to add locking around `self._control`, but could do that if you wish